### PR TITLE
enh/TSRI-98-migrate-to-other-payment-provider

### PR DESCRIPTION
**Pull request by michael-scheurer**
I decided to create a re-usable solution for the other media without the need of hard-coding in the front-end.

In the editor the media can set to migrate to a target payment method within a subscription plan (member plan).

This looks like this:
image

The migration hits as soon as the user calls a public subscription endpoint (e.g. extend a subscription or pay an open invoice etc.).

### CODE REVIEW

**Review by LukasSkywalker — APPROVED** [review]

File: libs/api/prisma/migrations/20241022171553_add_migrate_to_target_pm_to_member_plan/migration.sql [hunk 1]
```diff
@@ -0,0 +1,5 @@
+-- AlterTable
+ALTER TABLE "member.plans" ADD COLUMN     "migrateToTargetPMid" TEXT;
```
Comment (elias-summermatter) [inline libs/api/prisma/migrations/20241022171553_add_migrate_to_target_pm_to_member_plan/migration.sql:2]: Is TEXT here correct shouln't this be VARCHAR? ^^

File: libs/api/src/lib/graphql/mutation.public.ts [hunk 2]
```diff
@@ -581,8 +589,11 @@ export const GraphQLPublicMutation = new GraphQLObjectType<undefined, Context>({
         const subscription = (await prisma.subscription.findUnique({
           where: {
             id: subscriptionId
+          },
+          include: {
+            memberPlan: true
           }
-        })) as SubscriptionWithRelations
+        })) as SubscriptionWithRelations & {memberPlan: MemberPlan}
```
Comment (elias-summermatter) [inline libs/api/src/lib/graphql/mutation.public.ts:596]: Why do we do this explicit casting should prisma not return the type save type?

File: libs/api/src/lib/graphql/mutation.public.ts [hunk 3]
```diff
@@ -968,10 +981,11 @@ export const GraphQLPublicMutation = new GraphQLObjectType<undefined, Context>({
         return await createPaymentWithProvider({
           paymentMethodID: paymentMethod.id,
           invoice,
-          saveCustomer: false,
+          saveCustomer: true,
```
Comment (elias-summermatter) [inline libs/api/src/lib/graphql/mutation.public.ts:984]: Why was this changed?

File: libs/api/src/lib/graphql/mutation.public.ts [hunk 4]
```diff
@@ -1014,9 +1032,10 @@ export const GraphQLPublicMutation = new GraphQLObjectType<undefined, Context>({
         return await createPaymentWithProvider({
           paymentMethodID: invoice.subscription.paymentMethodID,
           invoice,
-          saveCustomer: false,
+          saveCustomer: true,
```
Comment (elias-summermatter) [inline libs/api/src/lib/graphql/mutation.public.ts:1035]: Same here why?

File: apps/api-example/schema-v1-admin.graphql [hunk 5]
```diff
@@ -1926,6 +1927,7 @@ input MemberPlanInput {
   extendable: Boolean!
   maxCount: Int
   availablePaymentMethods: [AvailablePaymentMethodInput!]!
+  migrateToTargetPMid: ID
```
Comment (elias-summermatter) [inline apps/api-example/schema-v1-admin.graphql:1930]: Should it not be ID= migrateToTargetPMID as it is for all others?

File: libs/api/__tests__/api/private/index.ts [hunk 6]
```diff
@@ -959,6 +959,7 @@ export type MemberPlan = {
   id: Scalars['ID'];
   image?: Maybe<Image>;
   maxCount?: Maybe<Scalars['Int']>;
+  migrateToTargetPMid?: Maybe<Scalars['ID']>;
```
Comment (elias-summermatter) [inline libs/api/__tests__/api/private/index.ts:962]: Why is here just the ID field compared to  availablePaymentMethods where the PaymentMethode is passed?

File: libs/api/prisma/schema.prisma [hunk 7]
```diff
@@ -509,6 +509,8 @@ model MemberPlan {
   currency                Currency
   amountPerMonthMin       Float
   availablePaymentMethods AvailablePaymentMethod[]
+  migrateToTargetPM       PaymentMethod?           @relation("PaymentMethodMigration", fields: [migrateToTargetPMid], references: [id])
+  migrateToTargetPMid     String?
```
Comment (elias-summermatter) [inline libs/api/prisma/schema.prisma:513]: Keep Naming consistent (ID)

File: libs/api/src/lib/context.ts [hunk 8]
```diff
@@ -225,6 +225,7 @@ export interface CreatePaymentWithProvider {
   successURL?: string
   failureURL?: string
   user?: User
+  migrateToTargetPMid?: string
```
Comment (elias-summermatter) [inline libs/api/src/lib/context.ts:228]: id=>ID

File: libs/api/src/lib/context.ts [hunk 9]
```diff
@@ -1003,9 +1004,12 @@ export async function contextFromRequest(
       saveCustomer,
       failureURL,
       successURL,
-      user
+      user,
+      migrateToTargetPMid
```
Comment (elias-summermatter) [inline libs/api/src/lib/context.ts:1008]: Why not accept here the subscription with memberpaln insted of some id? Like this we can determine the correct payment method from the subscription.

File: libs/api/src/lib/context.ts [hunk 10]
```diff
@@ -1014,6 +1018,26 @@ export async function contextFromRequest(
...
+        const updatedSubscription = await prisma.subscription.update({
+          data: {
+            paymentMethodID: migrateToTargetPMid
+          },
+          where: {
+            id: invoice.subscriptionID
+          }
+        })
+
+        if (!updatedSubscription) {
```
Comment (elias-summermatter) [inline libs/api/src/lib/context.ts:1036]: Since the update methode of prisma throws if update does not work that is not nesseary.

File: libs/api/src/lib/context.ts [hunk 11]
```diff
@@ -1014,6 +1018,26 @@ export async function contextFromRequest(
...
+       * Mainly used in mutation.public.ts
+       * Requirements written down here https://wepublish.atlassian.net/browse/TSRI-98
+       */
+      if (migrateToTargetPMid) {
+        const updatedSubscription = await prisma.subscription.update({
+          data: {
+            paymentMethodID: migrateToTargetPMid
+          },
+          where: {
+            id: invoice.subscriptionID
```
Comment (elias-summermatter) [inline libs/api/src/lib/context.ts:1032]: If you get the subscription passed you can use here subscription.id

File: libs/api/src/lib/graphql/mutation.public.ts [hunk 12]
```diff
@@ -421,7 +427,8 @@ export const GraphQLPublicMutation = new GraphQLObjectType<undefined, Context>({
             saveCustomer: true,
             paymentMethodID: paymentMethod.id,
             successURL,
-            failureURL
+            failureURL,
+            migrateToTargetPMid: memberPlan.migrateToTargetPMid
```
Comment (elias-summermatter) [inline libs/api/src/lib/graphql/mutation.public.ts:431]: Why is this needed in the registerMemberAndReceivePayment function should here not be the payment methode already correct since its a new subscription?

File: libs/api/src/lib/graphql/mutation.public.ts [hunk 13]
```diff
@@ -693,7 +704,8 @@ export const GraphQLPublicMutation = new GraphQLObjectType<undefined, Context>({
           saveCustomer: true,
           paymentMethodID: subscription.paymentMethodID,
           successURL,
-          failureURL
+          failureURL,
+          migrateToTargetPMid: subscription.memberPlan.migrateToTargetPMid
```
Comment (elias-summermatter) [inline libs/api/src/lib/graphql/mutation.public.ts:708]: In the  extendSubscription it is correct.

File: libs/api/src/lib/graphql/mutation.public.ts [hunk 14]
```diff
@@ -968,10 +981,11 @@ export const GraphQLPublicMutation = new GraphQLObjectType<undefined, Context>({
         return await createPaymentWithProvider({
           paymentMethodID: paymentMethod.id,
           invoice,
-          saveCustomer: false,
+          saveCustomer: true,
           successURL,
           failureURL,
-          user
+          user,
+          migrateToTargetPMid: subscription.memberPlan.migrateToTargetPMid
```
Comment (elias-summermatter) [inline libs/api/src/lib/graphql/mutation.public.ts:988]: createPaymentFromInvoice its fine too

File: libs/api/src/lib/graphql/mutation.public.ts [hunk 15]
```diff
@@ -1014,9 +1032,10 @@ export const GraphQLPublicMutation = new GraphQLObjectType<undefined, Context>({
         return await createPaymentWithProvider({
           paymentMethodID: invoice.subscription.paymentMethodID,
           invoice,
-          saveCustomer: false,
+          saveCustomer: true,
           successURL,
-          failureURL
+          failureURL,
+          migrateToTargetPMid: invoice.subscription.memberPlan.migrateToTargetPMid
```
Comment (elias-summermatter) [inline libs/api/src/lib/graphql/mutation.public.ts:1038]: createPaymentFromSubscription also required

**Review by elias-summermatter — CHANGES_REQUESTED** [review]

File: libs/api/src/lib/graphql/mutation.public.ts (hunk 3 above)
Comment (michael-scheurer) [reply libs/api/src/lib/graphql/mutation.public.ts:984]: So that the new credit card details would be stored

File: libs/api/src/lib/graphql/mutation.public.ts (hunk 4 above)
Comment (michael-scheurer) [reply libs/api/src/lib/graphql/mutation.public.ts:1035]: same answer as above ;9

File: libs/api/prisma/migrations/20241022171553_add_migrate_to_target_pm_to_member_plan/migration.sql (hunk 1 above)
Comment (michael-scheurer) [reply libs/api/prisma/migrations/20241022171553_add_migrate_to_target_pm_to_member_plan/migration.sql:2]: That's how prisma converts it. This is consistent to the rest of the schema. See schema.prisma

File: libs/api/src/lib/graphql/mutation.public.ts (hunk 2 above)
Comment (michael-scheurer) [reply libs/api/src/lib/graphql/mutation.public.ts:596]: Typescript forced me to add this line.

File: apps/api-example/schema-v1-admin.graphql (hunk 5 above)
Comment (michael-scheurer) [reply apps/api-example/schema-v1-admin.graphql]: changed it.

File: libs/api/__tests__/api/private/index.ts (hunk 6 above)
Comment (michael-scheurer) [reply libs/api/__tests__/api/private/index.ts]: the migration property lives on the member plan directly and not on a n:m available payment method relations. that is how available payment methods should also be related some day.

File: libs/api/src/lib/context.ts (hunk 8 above)
Comment (michael-scheurer) [reply libs/api/src/lib/context.ts]: done

File: libs/api/src/lib/context.ts (hunk 9 above)
Comment (michael-scheurer) [reply libs/api/src/lib/context.ts]: The idea is, that we can tell this method, whether to migrate or not. Would like to keep it that way.

File: libs/api/src/lib/context.ts (hunk 11 above)
Comment (michael-scheurer) [reply libs/api/src/lib/context.ts:1032]: idem

File: libs/api/src/lib/context.ts (hunk 10 above)
Comment (michael-scheurer) [reply libs/api/src/lib/context.ts:1036]: :heavy_check_mark: fixed

File: libs/api/src/lib/graphql/mutation.public.ts (hunk 12 above)
Comment (michael-scheurer) [reply libs/api/src/lib/graphql/mutation.public.ts]: one could argue so. probably the expected behavior is more concise. 

:heavy_check_mark: done

**Review by elias-summermatter — APPROVED** [review]
